### PR TITLE
Validate user specified arguments and populate fields in the created Performance Profile

### PR DESCRIPTION
**Pull request by swatisehgal**
In this PR we:

- Refactor the code to move the code in RunE responsible for obtaining profile data from flags and creating the performance profile into private methods that are then called from RunE.
- Validate and populate Topology Manager Policy and  Real Time Kernel flag in performance profile as specified by the user.
- Populate Real Time Kernel flag in performance profile as specified by the user.
- Validate power-consumption-mode flag value provided by user.
- Ensure that reservedCPUCount is less than total number of CPUs.

### CODE REVIEW

File: cmd/performance-profile-creator/cmd/root.go [hunk 1]
```diff
@@ -34,11 +34,20 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const (
+	singleNumaNodeTopologyManagerPolicy = "single-numa-node"
+)
+
+var (
+	validTMPolicyValues = []string{"single-numa-node", "best-effort", "restricted", "none", ""}
```
Comment (marcel-apf) [inline cmd/performance-profile-creator/cmd/root.go]: Why do we need 2 NOP options, "none" and  "" ?
Can we even have a none TM policy? I wasn't aware of it.

File: cmd/performance-profile-creator/cmd/root.go [hunk 2]
```diff
@@ -80,12 +89,24 @@ func getDataFromFlags(cmd *cobra.Command) (*profileCreatorArgs, error) {
 		return nil, fmt.Errorf("Error parsing split-reserved-cpus-across-numa flag: %v", err)
 	}
 	profileName := cmd.Flag("profile-name").Value.String()
+	tmPolicy := cmd.Flag("topology-manager-policy").Value.String()
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing topology-manager-policy flag: %v", err)
```
Comment (marcel-apf) [inline cmd/performance-profile-creator/cmd/root.go:94]: We will need to go over all error messages and remove the capital letters and even the "Error" prefix, but is low priority and not part of this PR.

File: cmd/performance-profile-creator/cmd/root.go [hunk 3]
```diff
@@ -39,7 +39,8 @@ const (
 )
 
 var (
-	validTMPolicyValues = []string{"single-numa-node", "best-effort", "restricted", "none", ""}
+	validTMPolicyValues        = []string{"single-numa-node", "best-effort", "restricted", "none", ""}
+	validPowerConsumptionModes = []string{"cstate", "no-state", "idle-poll"}
```
Comment (marcel-apf) [inline cmd/performance-profile-creator/cmd/root.go:41]: @MarSik @fromanirh are the names clear enough? 
What we have in design is:
Power consumption mode
   Default (cstate/pstate sleep level limit to 1)
   Ignore power consumption (don’t use c-states to save power)
   No power saving  - (idle=pool))

Maybe we can have
   "limit-cstates" - default
   "ignore"
    "no-power-saving"

Can you come up with better names?

Also do you remember maybe what kernel args *exactly* we should use for it? We can delay this part for when we'll send a PR for it, but maybe you have the answer.

File: cmd/performance-profile-creator/cmd/root.go (hunk 1 above)
Comment (swatisehgal) [reply cmd/performance-profile-creator/cmd/root.go]: We don't need "" anymore. In the previous implementation, in case the user wasn't specifying any value it would default to "" but as we default to "restricted" that is no longer relevant.
Yes "none" is one the TM policies. Please refer [here](https://kubernetes.io/docs/tasks/administer-cluster/topology-manager/#topology-manager-policies).

File: cmd/performance-profile-creator/cmd/root.go (hunk 2 above)
Comment (swatisehgal) [reply cmd/performance-profile-creator/cmd/root.go:94]: No problem, can fix this.

File: cmd/performance-profile-creator/cmd/root.go (hunk 3 above)
Comment (fromanirh) [reply cmd/performance-profile-creator/cmd/root.go:41]: My take: I'd abstract the power saving setting from the implementation, hence I'd go with something like

- default: maps to your default
- performance: maps to your "ignore power consumption" (is it actually ignoring?)
- low-latency: maps to "no power saving"

or something like it. The docs should also further clarify that the higher we go (low-latency is higher than performance) the more power the system will consume.

File: cmd/performance-profile-creator/cmd/root.go [hunk 4]
```diff
@@ -38,67 +38,23 @@ import (
...
 
 // rootCmd represents the base command when called without any subcommands
 var rootCmd = &cobra.Command{
 	Use:   "performance-profile-creator",
 	Short: "A tool that automates creation of Performance Profiles",
 	RunE: func(cmd *cobra.Command, args []string) error {
-		mcpName := cmd.Flag("mcp-name").Value.String()
-		mustGatherDirPath := cmd.Flag("must-gather-dir-path").Value.String()
-		mcp, err := profilecreator.GetMCP(mustGatherDirPath, mcpName)
+		flagData, err := getDataFromFlags(cmd)
```
Comment (fromanirh) [inline cmd/performance-profile-creator/cmd/root.go:58]: I'd rename like `profileCreatorArgsFromFlags`
Additionally: we can return a value (not a pointer) here, can't we?

File: cmd/performance-profile-creator/cmd/root.go [hunk 5]
```diff
@@ -38,67 +38,23 @@ import (
...
-		handle, err := profilecreator.NewGHWHandler(mustGatherDirPath, matchedNodes[0])
-		reservedCPUs, isolatedCPUs, err := handle.GetReservedAndIsolatedCPUs(reservedCPUCount, splitReservedCPUsAcrossNUMA)
-		if err != nil {
-			return fmt.Errorf("Error obtaining Reserved and Isolated CPUs for %s: %v", nodeName, err)
-		}
-		profileData = ProfileData{
-			reservedCPUs: reservedCPUs,
-			isolatedCPUs: isolatedCPUs,
-			nodeSelector: mcp.Spec.NodeSelector,
+			return fmt.Errorf("Error obtaining profileData %v", err)
```
Comment (fromanirh) [inline cmd/performance-profile-creator/cmd/root.go]: I'd use `error creating the profile: %v`

File: cmd/performance-profile-creator/cmd/root.go (hunk 1 above)
Comment (fromanirh) [reply cmd/performance-profile-creator/cmd/root.go]: can't we just import the k8s/ocp names? duplicating data (and validation) always carries some risk.

File: cmd/performance-profile-creator/cmd/root.go [hunk 6]
```diff
@@ -131,10 +152,23 @@ func getProfileData(args *profileCreatorArgs) (*ProfileData, error) {
...
 		nodeSelector:           mcp.Spec.NodeSelector,
 		performanceProfileName: args.profileName,
+		topologyPoilcy:         args.tmPolicy,
 	}
 	return profileData, nil
 }
 
+func validateFlag(value string, validValues []string) error {
+	for _, validValue := range validValues {
+		if strings.ToLower(value) == validValue {
```
Comment (fromanirh) [inline cmd/performance-profile-creator/cmd/root.go:177]: `strings.EqualFold` (https://golang.org/pkg/strings/#EqualFold) will probably come handy here

Comment (fromanirh) [reply cmd/performance-profile-creator/cmd/root.go:177]: the string matching should be extracted:

```
func validateFlag(value string, validValues []string) error {
	if isStringInSlice(value, validValues) {
		return nil
       }
	return fmt.Errorf("Value '%s' is invalid for flag 'color'. Valid values "+
		"come from the set %v", value, validValues)
}

func isStringInSlice(value string, candidates []string) bool {
	for _, candidate := range candidates {
		if strings.EqualFold(candidate, value) {
			return true
		}
	}
	return false
}
```

File: cmd/performance-profile-creator/cmd/root.go [hunk 7]
```diff
@@ -163,6 +198,7 @@ func init() {
 	rootCmd.PersistentFlags().StringVarP(&args.mustGatherDirPath, "must-gather-dir-path", "M", "must-gather", "Must gather directory path")
 	rootCmd.MarkPersistentFlagRequired("must-gather-dir-path")
 	rootCmd.PersistentFlags().StringVarP(&args.profileName, "profile-name", "N", "performance", "Name of the performance profile to be created")
+	rootCmd.PersistentFlags().StringVarP(&args.tmPolicy, "topology-manager-policy", "Q", "restricted", "Kubelet Topology Manager Policy of the performance profile to be created. [Valid values: single-numa-node, best-effort, restricted, none]")
```
Comment (fromanirh) [inline cmd/performance-profile-creator/cmd/root.go:216]: why "Q"?

File: pkg/profilecreator/profilecreator.go [hunk 8]
```diff
@@ -254,8 +254,12 @@ func (ghwHandler GHWHandler) SortedTopology() (*topology.Info, error) {
...
 // GetReservedAndIsolatedCPUs returns Reserved and Isolated CPUs
 func (ghwHandler GHWHandler) GetReservedAndIsolatedCPUs(reservedCPUCount int, splitReservedCPUsAcrossNUMA bool) (string, string, error) {
-	if reservedCPUCount < 0 {
-		return "", "", fmt.Errorf("Specified eservered CPU count is negative, please specify it correctly")
+	cpuInfo, err := ghwHandler.CPU()
+	if err != nil {
+		return "", "", fmt.Errorf("Error obtaining CPU Info from GHW snapshot: %v", err)
+	}
+	if reservedCPUCount < 0 || reservedCPUCount > int(cpuInfo.TotalThreads) {
+		return "", "", fmt.Errorf("Specified reserved CPU count is invalid, please specify it correctly")
```
Comment (fromanirh) [inline pkg/profilecreator/profilecreator.go]: this is very nicely worded but not helpful. Let's tell the user what PPC thinks is correct:
```
fmt.Errorf("Specified reserved CPU count is invalid, please specify it in the range [0,%d]", cpuInfo.TotalThreads)

File: cmd/performance-profile-creator/cmd/root.go (hunk 7 above)
Comment (swatisehgal) [reply cmd/performance-profile-creator/cmd/root.go:216]: Ran out of letter :) We used up T, M and P so thought Q for Kubelet (Qbelet) Topology Manager Policy purely for the phonetic reasons

File: cmd/performance-profile-creator/cmd/root.go (hunk 6 above)
Comment (swatisehgal) [reply cmd/performance-profile-creator/cmd/root.go:177]: Thanks, will fix

File: pkg/profilecreator/profilecreator.go (hunk 8 above)
Comment (swatisehgal) [reply pkg/profilecreator/profilecreator.go]: +1, will do!

File: cmd/performance-profile-creator/cmd/root.go [hunk 9]
```diff
@@ -30,15 +30,16 @@ import (
...
 )
 
 var (
-	validTMPolicyValues        = []string{"single-numa-node", "best-effort", "restricted", "none"}
-	validPowerConsumptionModes = []string{"cstate", "no-state", "idle-poll"}
+	validTMPolicyValues = []string{kubeletconfig.SingleNumaNodeTopologyManager, kubeletconfig.BestEffortTopologyManagerPolicy, kubeletconfig.RestrictedTopologyManagerPolicy}
+	// Values part of validPowerConsumptionModes are explained below:
+	// default => Use CPU C-states but limit them to 1, meaning the processor is idle it can sleep but not "deep sleep"
+	// performance => Disable CPU sleep (c-states), processor never sleeps even if is idle
+	// low-latency => processor is never idle, it is in polling mode (cpu=poll)
```
Comment (marcel-apf) [inline cmd/performance-profile-creator/cmd/root.go:41]: performance + idle=poll (not cpu=poll, my mistake)

File: cmd/performance-profile-creator/cmd/root.go [hunk 10]
```diff
@@ -30,15 +30,16 @@ import (
...
+	kubeletconfig "k8s.io/kubelet/config/v1beta1"
 )
 
 var (
-	validTMPolicyValues        = []string{"single-numa-node", "best-effort", "restricted", "none"}
-	validPowerConsumptionModes = []string{"cstate", "no-state", "idle-poll"}
+	validTMPolicyValues = []string{kubeletconfig.SingleNumaNodeTopologyManager, kubeletconfig.BestEffortTopologyManagerPolicy, kubeletconfig.RestrictedTopologyManagerPolicy}
+	// Values part of validPowerConsumptionModes are explained below:
+	// default => Use CPU C-states but limit them to 1, meaning the processor is idle it can sleep but not "deep sleep"
+	// performance => Disable CPU sleep (c-states), processor never sleeps even if is idle
```
Comment (marcel-apf) [inline cmd/performance-profile-creator/cmd/root.go:40]: intel_idle.max_cstate=0 processor.max_cstate=0 intel_pstate=disable

File: cmd/performance-profile-creator/cmd/root.go [hunk 11]
```diff
@@ -30,15 +30,16 @@ import (
...
-	singleNumaNodeTopologyManagerPolicy = "single-numa-node"
+	kubeletconfig "k8s.io/kubelet/config/v1beta1"
 )
 
 var (
-	validTMPolicyValues        = []string{"single-numa-node", "best-effort", "restricted", "none"}
-	validPowerConsumptionModes = []string{"cstate", "no-state", "idle-poll"}
+	validTMPolicyValues = []string{kubeletconfig.SingleNumaNodeTopologyManager, kubeletconfig.BestEffortTopologyManagerPolicy, kubeletconfig.RestrictedTopologyManagerPolicy}
+	// Values part of validPowerConsumptionModes are explained below:
+	// default => Use CPU C-states but limit them to 1, meaning the processor is idle it can sleep but not "deep sleep"
```
Comment (marcel-apf) [inline cmd/performance-profile-creator/cmd/root.go:39]: "processor.max_cstate=1"

File: cmd/performance-profile-creator/cmd/root.go (hunk 9 above)
Comment (marcel-apf) [reply cmd/performance-profile-creator/cmd/root.go:41]: @MarSik can you please confirm the kernel args for each power mode? Thank you

File: cmd/performance-profile-creator/cmd/root.go [hunk 12]
```diff
@@ -203,17 +210,17 @@ func init() {
 	rootCmd.PersistentFlags().IntVarP(&args.reservedCPUCount, "reserved-cpu-count", "R", 0, "Number of reserved CPUs (required)")
 	rootCmd.MarkPersistentFlagRequired("reserved-cpu-count")
 	rootCmd.PersistentFlags().BoolVarP(&args.splitReservedCPUsAcrossNUMA, "split-reserved-cpus-across-numa", "S", false, "Split the Reserved CPUs across NUMA nodes")
-	rootCmd.PersistentFlags().StringVarP(&args.mcpName, "mcp-name", "T", "worker-cnf", "MCP name corresponding to the target machines (required)")
+	rootCmd.PersistentFlags().StringVarP(&args.mcpName, "mcp-name", "n", "worker-cnf", "MCP name corresponding to the target machines (required)")
```
Comment (marcel-apf) [inline cmd/performance-profile-creator/cmd/root.go:213]: Why not capital letter? it may confuse

File: pkg/profilecreator/profilecreator.go [hunk 13]
```diff
@@ -259,7 +259,7 @@ func (ghwHandler GHWHandler) GetReservedAndIsolatedCPUs(reservedCPUCount int, sp
 		return "", "", fmt.Errorf("can't obtain CPU info from GHW snapshot: %v", err)
 	}
 	if reservedCPUCount < 0 || reservedCPUCount > int(cpuInfo.TotalThreads) {
```
Comment (marcel-apf) [inline pkg/profilecreator/profilecreator.go:261]: is  reservedCPUCount == int(cpuInfo.TotalThreads)  a valid choice? Not a single cpu for workloads? We may want to error out or to issue a warning at least

File: cmd/performance-profile-creator/cmd/root.go (hunk 9 above)
Comment (MarSik) [reply cmd/performance-profile-creator/cmd/root.go:41]: I wonder what is the difference between performance and low-latency. It is basically the same just using a different method, isn't it?

File: cmd/performance-profile-creator/cmd/root.go (hunk 10 above)
Comment (MarSik) [reply cmd/performance-profile-creator/cmd/root.go:40]: - "processor.max_cstate=1" <- This does not support 0 internally, it converts it to 1 anyway
    - "intel_idle.max_cstate=0"

File: cmd/performance-profile-creator/cmd/root.go (hunk 12 above)
Comment (swatisehgal) [reply cmd/performance-profile-creator/cmd/root.go:213]: I made updates based on @fromanirh suggestion [here](https://github.com/openshift-kni/performance-addon-operators/pull/570#discussion_r590338740).

In the previous patch, we used shorthand for topology-manager-policy flag as "Q" (discussion about this [here](https://github.com/openshift-kni/performance-addon-operators/pull/570#discussion_r590322813)) and updated shorthand for mcp-name to "n" (n for name as "N" was already used for policy-name) so I could use "T" for topology-manager-policy.

Please let me know if that is okay? If not please let me know what shorthand letter you would like me to use for topology-manager-policy.

Comment (marcel-apf) [reply cmd/performance-profile-creator/cmd/root.go:213]: I am "lost" a little in letters. I have no special "preference" as long as all are Capital letters (or not)

Comment (fromanirh) [reply cmd/performance-profile-creator/cmd/root.go:213]: the best option is just to get rid of single-letter options.

Comment (swatisehgal) [reply cmd/performance-profile-creator/cmd/root.go:213]: Sure, can certainly do that if it causes less confusion :)

**Review by fromanirh — COMMENTED** [review]
/lgtm
/hold
thanks for following through. All my comments are addressed so I'm happy with the state of the current PR.
Please remove the hold when the other reviewers are happy.

**Comment by marcel-apf** [comment]
Looks good, thank you!
/unhold
